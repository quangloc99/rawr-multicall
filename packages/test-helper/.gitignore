artifacts
ethers-v6-contracts/typechain-types