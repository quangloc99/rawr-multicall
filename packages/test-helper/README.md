# @rawr-multicall/test-helper

Test helpers for [Rawr-multicall].

Please refer to [Rawr-mulitcall] for more details.

[Rawr-multicall]: https://github.com/quangloc99/rawr-multicall
