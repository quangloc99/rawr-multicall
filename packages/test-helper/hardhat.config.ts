import '@typechain/hardhat';
import { HardhatUserConfig } from 'hardhat/config';

const config: HardhatUserConfig = {
    paths: {
        sources: 'test-contracts',
        artifacts: 'artifacts',
    },
    typechain: {
        outDir: 'typechain-types-ethers-v6',
    },
    solidity: {
        version: '0.8.17',
    },
};

export default config;
