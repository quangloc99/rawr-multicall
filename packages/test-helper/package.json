{
    "name": "@raw-multicall/test-helper",
    "version": "1.0.0",
    "packageManager": "yarn@3.3.1",
    "main": "dist/index.js",
    "types": "dist/index.d.js",
    "files": [
        "dist",
        "src",
        "artifacts"
    ],
    "scripts": {
        "typecheck": "yarn tsc --noEmit",
        "build": "yarn tsc -p tsconfig.build.json",
        "lint:nofix": "yarn eslint \"{src,test}/**/*.{ts,js}\" --quiet",
        "lint": "yarn lint:nofix --fix",
        "compile-contracts-ethers-v6": "yarn hardhat compile",
        "compile-contracts": "yarn compile-contracts-ethers-v6 --force",
        "prepack": "yarn build && yarn compile-contracts"
    },
    "devDependencies": {
        "@typechain/ethers-v6": "^0.4.0",
        "@typechain/hardhat": "^8.0.0",
        "@typescript-eslint/eslint-plugin": "^6.0.0",
        "@typescript-eslint/parser": "^6.0.0",
        "eslint": "^8.44.0",
        "eslint-config-prettier": "^8.8.0",
        "eslint-plugin-prettier": "^5.0.0",
        "eslint-plugin-unused-imports": "^3.0.0",
        "ethers": "^6.6.3",
        "hardhat": "^2.16.1",
        "prettier": "^3.0.0",
        "typechain": "^8.2.0",
        "typescript": "^5.1.6"
    }
}
