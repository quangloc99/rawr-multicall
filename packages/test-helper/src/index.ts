import { CHAIN_ID_MAPPING } from './chainId';
import { testEnv } from './test-env';
import { RPC_URL } from './rpc-url';

export { testEnv, RPC_URL, CHAIN_ID_MAPPING };
export * from './jest.config';
export * as testData from './test-data';

export function describeForChain(
    ...params:
        | [callback: (rpcUrl: string, chain: number) => void]
        | [nameFormatter: (chain: number) => string, callback: (rpcUrl: string, chain: number) => void]
) {
    const [nameFormatter, callback] =
        params.length == 2 ? params : [(chain: number) => `Testing chain ${chain}`, ...params];

    for (const chain of Object.values(CHAIN_ID_MAPPING)) {
        const isAllowedChain = testEnv.CHAIN_ID == undefined || chain === testEnv.CHAIN_ID;
        const name = nameFormatter(chain);
        (isAllowedChain ? describe : describe.skip)(name, () => callback(RPC_URL[chain], chain));
    }
}
