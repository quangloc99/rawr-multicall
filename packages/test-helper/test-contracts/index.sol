import './aPlusB.sol';
import './ERC20.sol';
import './TestContract.sol';
import './ThrowError.sol';
