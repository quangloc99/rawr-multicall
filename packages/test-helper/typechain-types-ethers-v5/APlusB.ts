/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BigNumberish,
  BytesLike,
  CallOverrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
} from "./common";

export interface APlusBInterface extends utils.Interface {
  functions: {
    "minus(int256,int256)": FunctionFragment;
    "plus(int256,int256)": FunctionFragment;
  };

  getFunction(nameOrSignatureOrTopic: "minus" | "plus"): FunctionFragment;

  encodeFunctionData(
    functionFragment: "minus",
    values: [BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "plus",
    values: [BigNumberish, BigNumberish]
  ): string;

  decodeFunctionResult(functionFragment: "minus", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "plus", data: BytesLike): Result;

  events: {};
}

export interface APlusB extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: APlusBInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    minus(
      a: BigNumberish,
      b: BigNumberish,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;

    plus(
      a: BigNumberish,
      b: BigNumberish,
      overrides?: CallOverrides
    ): Promise<[BigNumber]>;
  };

  minus(
    a: BigNumberish,
    b: BigNumberish,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  plus(
    a: BigNumberish,
    b: BigNumberish,
    overrides?: CallOverrides
  ): Promise<BigNumber>;

  callStatic: {
    minus(
      a: BigNumberish,
      b: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    plus(
      a: BigNumberish,
      b: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  filters: {};

  estimateGas: {
    minus(
      a: BigNumberish,
      b: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;

    plus(
      a: BigNumberish,
      b: BigNumberish,
      overrides?: CallOverrides
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    minus(
      a: BigNumberish,
      b: BigNumberish,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    plus(
      a: BigNumberish,
      b: BigNumberish,
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;
  };
}
