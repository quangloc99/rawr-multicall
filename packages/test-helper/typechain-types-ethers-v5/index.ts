/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
export type { ERC20 } from "./ERC20";
export type { TestContract } from "./TestContract";
export type { ThrowError } from "./ThrowError";
export type { APlusB } from "./APlusB";
export * as factories from "./factories";
export { APlusB__factory } from "./factories/APlusB__factory";
export { ERC20__factory } from "./factories/ERC20__factory";
export { TestContract__factory } from "./factories/TestContract__factory";
export { ThrowError__factory } from "./factories/ThrowError__factory";
