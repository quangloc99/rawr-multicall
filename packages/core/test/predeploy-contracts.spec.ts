import { describeForChain, CHAIN_ID_MAPPING } from '@raw-multicall/test-helper';
import { ethers } from 'ethers';
import { APlusB__factory } from '@raw-multicall/test-helper/typechain-types-ethers-v6';
import { APlusBInterface } from '@raw-multicall/test-helper/typechain-types-ethers-v6/APlusB';
import {
    labeledAddress,
    buildRawMulticallContract,
    createCall,
    decodeRawResult,
    NoPredeployContractError,
    resetPredeployContracts,
    registerPredeployContract,
    strip0x,
} from '../src';

describeForChain(
    (chain: number) => `Test predeploy contracts for chain ${chain}`,
    (rpcUrl, chain) => {
        const provider = new ethers.JsonRpcProvider(rpcUrl);
        const APlusBFactory = new APlusB__factory();
        const APlusBIface = new ethers.Interface(APlusB__factory.abi) as unknown as APlusBInterface;
        const allowPUSH0 = chain !== CHAIN_ID_MAPPING.ARBITRUM;

        beforeEach(() => {
            resetPredeployContracts();
        });

        it('simple', async () => {
            const calls = [
                createCall(labeledAddress('testContract'), APlusBIface.encodeFunctionData('plus', [10, 20])),
                createCall(labeledAddress('testContract'), APlusBIface.encodeFunctionData('minus', [10, 20])),
            ];
            const testContract = (await APlusBFactory.getDeployTransaction()).data;
            const callData = buildRawMulticallContract(calls, {
                allowPUSH0,
                predeployContracts: { testContract },
            });
            expect(callData).toMatchSnapshot();
            const res = await provider.call({ data: callData.byteCode });
            expect(res).toMatchSnapshot();
            expect(callData.byteCode.includes(strip0x(testContract)));
            const result = decodeRawResult(res);
            expect(result).toMatchSnapshot();
            expect(APlusBIface.decodeFunctionResult('plus', result[0].data)).toMatchSnapshot();
            expect(APlusBIface.decodeFunctionResult('minus', result[1].data)).toMatchSnapshot();
        });

        it('no predeployed', () => {
            registerPredeployContract('existing-contract', '0x');
            const calls = [createCall(labeledAddress('non-existing'), '0x')];
            expect(() => buildRawMulticallContract(calls)).toThrowError(NoPredeployContractError);
        });

        it('register predeploy contract', async () => {
            const calls = [
                createCall(labeledAddress('my-contract'), APlusBIface.encodeFunctionData('plus', [69, 420])),
            ];
            const myContract = (await APlusBFactory.getDeployTransaction()).data;
            registerPredeployContract('my-contract', myContract);
            const callData = buildRawMulticallContract(calls, {
                allowPUSH0,
            });
            expect(callData).toMatchSnapshot();
            expect(callData.byteCode.includes(strip0x(myContract)));
            const res = await provider.call({ data: callData.byteCode });
            expect(res).toMatchSnapshot();
            const result = decodeRawResult(res);
            expect(result).toMatchSnapshot();
        });
    }
);
