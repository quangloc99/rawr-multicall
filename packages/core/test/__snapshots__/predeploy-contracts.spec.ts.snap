// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test predeploy contracts for chain 1 register predeploy contract 1`] = `
{
  "byteCode": "0x6102bc6200004560803961027860805ff061033c5260ff60e0600461035c5f5f60446102f85f61033c515af1841b3d841b01815281013d5f823e3d0161035c900361035cf3608060405234801561001057600080fd5b50610258806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c80631858c77d1461003b578063ccf655031461006b575b600080fd5b61005560048036038101906100509190610102565b61009b565b6040516100629190610151565b60405180910390f35b61008560048036038101906100809190610102565b6100b1565b6040516100929190610151565b60405180910390f35b600081836100a9919061019b565b905092915050565b600081836100bf91906101de565b905092915050565b600080fd5b6000819050919050565b6100df816100cc565b81146100ea57600080fd5b50565b6000813590506100fc816100d6565b92915050565b60008060408385031215610119576101186100c7565b5b6000610127858286016100ed565b9250506020610138858286016100ed565b9150509250929050565b61014b816100cc565b82525050565b60006020820190506101666000830184610142565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101a6826100cc565b91506101b1836100cc565b92508282039050818112600084121682821360008512151617156101d8576101d761016c565b5b92915050565b60006101e9826100cc565b91506101f4836100cc565b92508282019050828112156000831216838212600084121516171561021c5761021b61016c565b5b9291505056fea2646970667358221220f5b5af1db90e0f8d8bff4bcfa08f119f2227ca5d7d2e3ba623523e219c33728264736f6c63430008110033ccf65503000000000000000000000000000000000000000000000000000000000000004500000000000000000000000000000000000000000000000000000000000001a4",
  "splittedByteCodes": [
    "6102bc",
    "62000045",
    "6080",
    "39",
    "610278",
    "6080",
    "5f",
    "f0",
    "61033c52",
    "60ff",
    "60e0",
    "6004",
    "61035c",
    "5f",
    "5f",
    "6044",
    "6102f8",
    "5f",
    "61033c51",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "61035c",
    "90",
    "03",
    "61035c",
    "f3",
    "",
    "608060405234801561001057600080fd5b50610258806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c80631858c77d1461003b578063ccf655031461006b575b600080fd5b61005560048036038101906100509190610102565b61009b565b6040516100629190610151565b60405180910390f35b61008560048036038101906100809190610102565b6100b1565b6040516100929190610151565b60405180910390f35b600081836100a9919061019b565b905092915050565b600081836100bf91906101de565b905092915050565b600080fd5b6000819050919050565b6100df816100cc565b81146100ea57600080fd5b50565b6000813590506100fc816100d6565b92915050565b60008060408385031215610119576101186100c7565b5b6000610127858286016100ed565b9250506020610138858286016100ed565b9150509250929050565b61014b816100cc565b82525050565b60006020820190506101666000830184610142565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101a6826100cc565b91506101b1836100cc565b92508282039050818112600084121682821360008512151617156101d8576101d761016c565b5b92915050565b60006101e9826100cc565b91506101f4836100cc565b92508282019050828112156000831216838212600084121516171561021c5761021b61016c565b5b9291505056fea2646970667358221220f5b5af1db90e0f8d8bff4bcfa08f119f2227ca5d7d2e3ba623523e219c33728264736f6c63430008110033",
    "ccf65503000000000000000000000000000000000000000000000000000000000000004500000000000000000000000000000000000000000000000000000000000001a4",
  ],
}
`;

exports[`Test predeploy contracts for chain 1 register predeploy contract 2`] = `"0x8000002000000000000000000000000000000000000000000000000000000000000001e9"`;

exports[`Test predeploy contracts for chain 1 register predeploy contract 3`] = `
[
  {
    "data": "0x00000000000000000000000000000000000000000000000000000000000001e9",
    "success": true,
  },
]
`;

exports[`Test predeploy contracts for chain 1 simple 1`] = `
{
  "byteCode": "0x6103006200006360803961027860805ff06103805260ff60e060046103a05f5f60446102f85f610380515af1841b3d841b01815281013d5f823e3d015f5f604461033c5f610380515af1841b3d841b01815281013d5f823e3d016103a090036103a0f3608060405234801561001057600080fd5b50610258806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c80631858c77d1461003b578063ccf655031461006b575b600080fd5b61005560048036038101906100509190610102565b61009b565b6040516100629190610151565b60405180910390f35b61008560048036038101906100809190610102565b6100b1565b6040516100929190610151565b60405180910390f35b600081836100a9919061019b565b905092915050565b600081836100bf91906101de565b905092915050565b600080fd5b6000819050919050565b6100df816100cc565b81146100ea57600080fd5b50565b6000813590506100fc816100d6565b92915050565b60008060408385031215610119576101186100c7565b5b6000610127858286016100ed565b9250506020610138858286016100ed565b9150509250929050565b61014b816100cc565b82525050565b60006020820190506101666000830184610142565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101a6826100cc565b91506101b1836100cc565b92508282039050818112600084121682821360008512151617156101d8576101d761016c565b5b92915050565b60006101e9826100cc565b91506101f4836100cc565b92508282019050828112156000831216838212600084121516171561021c5761021b61016c565b5b9291505056fea2646970667358221220f5b5af1db90e0f8d8bff4bcfa08f119f2227ca5d7d2e3ba623523e219c33728264736f6c63430008110033ccf65503000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000141858c77d000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000014",
  "splittedByteCodes": [
    "610300",
    "62000063",
    "6080",
    "39",
    "610278",
    "6080",
    "5f",
    "f0",
    "61038052",
    "60ff",
    "60e0",
    "6004",
    "6103a0",
    "5f",
    "5f",
    "6044",
    "6102f8",
    "5f",
    "61038051",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "5f",
    "5f",
    "6044",
    "61033c",
    "5f",
    "61038051",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "5f",
    "82",
    "3e",
    "3d",
    "01",
    "6103a0",
    "90",
    "03",
    "6103a0",
    "f3",
    "",
    "608060405234801561001057600080fd5b50610258806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c80631858c77d1461003b578063ccf655031461006b575b600080fd5b61005560048036038101906100509190610102565b61009b565b6040516100629190610151565b60405180910390f35b61008560048036038101906100809190610102565b6100b1565b6040516100929190610151565b60405180910390f35b600081836100a9919061019b565b905092915050565b600081836100bf91906101de565b905092915050565b600080fd5b6000819050919050565b6100df816100cc565b81146100ea57600080fd5b50565b6000813590506100fc816100d6565b92915050565b60008060408385031215610119576101186100c7565b5b6000610127858286016100ed565b9250506020610138858286016100ed565b9150509250929050565b61014b816100cc565b82525050565b60006020820190506101666000830184610142565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101a6826100cc565b91506101b1836100cc565b92508282039050818112600084121682821360008512151617156101d8576101d761016c565b5b92915050565b60006101e9826100cc565b91506101f4836100cc565b92508282019050828112156000831216838212600084121516171561021c5761021b61016c565b5b9291505056fea2646970667358221220f5b5af1db90e0f8d8bff4bcfa08f119f2227ca5d7d2e3ba623523e219c33728264736f6c63430008110033",
    "ccf65503000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000141858c77d000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000014",
  ],
}
`;

exports[`Test predeploy contracts for chain 1 simple 2`] = `"0x80000020000000000000000000000000000000000000000000000000000000000000001e80000020fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff6"`;

exports[`Test predeploy contracts for chain 1 simple 3`] = `
[
  {
    "data": "0x000000000000000000000000000000000000000000000000000000000000001e",
    "success": true,
  },
  {
    "data": "0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff6",
    "success": true,
  },
]
`;

exports[`Test predeploy contracts for chain 1 simple 4`] = `
 [
  30n,
]
`;

exports[`Test predeploy contracts for chain 1 simple 5`] = `
 [
  -10n,
]
`;

exports[`Test predeploy contracts for chain 42161 register predeploy contract 1`] = `
{
  "byteCode": "0x6102bc6200004a60803961027860806000f061033c5260ff60e0600461035c6000600060446102f8600061033c515af1841b3d841b01815281013d6000823e3d0161035c900361035cf3608060405234801561001057600080fd5b50610258806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c80631858c77d1461003b578063ccf655031461006b575b600080fd5b61005560048036038101906100509190610102565b61009b565b6040516100629190610151565b60405180910390f35b61008560048036038101906100809190610102565b6100b1565b6040516100929190610151565b60405180910390f35b600081836100a9919061019b565b905092915050565b600081836100bf91906101de565b905092915050565b600080fd5b6000819050919050565b6100df816100cc565b81146100ea57600080fd5b50565b6000813590506100fc816100d6565b92915050565b60008060408385031215610119576101186100c7565b5b6000610127858286016100ed565b9250506020610138858286016100ed565b9150509250929050565b61014b816100cc565b82525050565b60006020820190506101666000830184610142565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101a6826100cc565b91506101b1836100cc565b92508282039050818112600084121682821360008512151617156101d8576101d761016c565b5b92915050565b60006101e9826100cc565b91506101f4836100cc565b92508282019050828112156000831216838212600084121516171561021c5761021b61016c565b5b9291505056fea2646970667358221220f5b5af1db90e0f8d8bff4bcfa08f119f2227ca5d7d2e3ba623523e219c33728264736f6c63430008110033ccf65503000000000000000000000000000000000000000000000000000000000000004500000000000000000000000000000000000000000000000000000000000001a4",
  "splittedByteCodes": [
    "6102bc",
    "6200004a",
    "6080",
    "39",
    "610278",
    "6080",
    "6000",
    "f0",
    "61033c52",
    "60ff",
    "60e0",
    "6004",
    "61035c",
    "6000",
    "6000",
    "6044",
    "6102f8",
    "6000",
    "61033c51",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "61035c",
    "90",
    "03",
    "61035c",
    "f3",
    "",
    "608060405234801561001057600080fd5b50610258806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c80631858c77d1461003b578063ccf655031461006b575b600080fd5b61005560048036038101906100509190610102565b61009b565b6040516100629190610151565b60405180910390f35b61008560048036038101906100809190610102565b6100b1565b6040516100929190610151565b60405180910390f35b600081836100a9919061019b565b905092915050565b600081836100bf91906101de565b905092915050565b600080fd5b6000819050919050565b6100df816100cc565b81146100ea57600080fd5b50565b6000813590506100fc816100d6565b92915050565b60008060408385031215610119576101186100c7565b5b6000610127858286016100ed565b9250506020610138858286016100ed565b9150509250929050565b61014b816100cc565b82525050565b60006020820190506101666000830184610142565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101a6826100cc565b91506101b1836100cc565b92508282039050818112600084121682821360008512151617156101d8576101d761016c565b5b92915050565b60006101e9826100cc565b91506101f4836100cc565b92508282019050828112156000831216838212600084121516171561021c5761021b61016c565b5b9291505056fea2646970667358221220f5b5af1db90e0f8d8bff4bcfa08f119f2227ca5d7d2e3ba623523e219c33728264736f6c63430008110033",
    "ccf65503000000000000000000000000000000000000000000000000000000000000004500000000000000000000000000000000000000000000000000000000000001a4",
  ],
}
`;

exports[`Test predeploy contracts for chain 42161 register predeploy contract 2`] = `"0x8000002000000000000000000000000000000000000000000000000000000000000001e9"`;

exports[`Test predeploy contracts for chain 42161 register predeploy contract 3`] = `
[
  {
    "data": "0x00000000000000000000000000000000000000000000000000000000000001e9",
    "success": true,
  },
]
`;

exports[`Test predeploy contracts for chain 42161 simple 1`] = `
{
  "byteCode": "0x6103006200006c60803961027860806000f06103805260ff60e060046103a06000600060446102f86000610380515af1841b3d841b01815281013d6000823e3d0160006000604461033c6000610380515af1841b3d841b01815281013d6000823e3d016103a090036103a0f3608060405234801561001057600080fd5b50610258806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c80631858c77d1461003b578063ccf655031461006b575b600080fd5b61005560048036038101906100509190610102565b61009b565b6040516100629190610151565b60405180910390f35b61008560048036038101906100809190610102565b6100b1565b6040516100929190610151565b60405180910390f35b600081836100a9919061019b565b905092915050565b600081836100bf91906101de565b905092915050565b600080fd5b6000819050919050565b6100df816100cc565b81146100ea57600080fd5b50565b6000813590506100fc816100d6565b92915050565b60008060408385031215610119576101186100c7565b5b6000610127858286016100ed565b9250506020610138858286016100ed565b9150509250929050565b61014b816100cc565b82525050565b60006020820190506101666000830184610142565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101a6826100cc565b91506101b1836100cc565b92508282039050818112600084121682821360008512151617156101d8576101d761016c565b5b92915050565b60006101e9826100cc565b91506101f4836100cc565b92508282019050828112156000831216838212600084121516171561021c5761021b61016c565b5b9291505056fea2646970667358221220f5b5af1db90e0f8d8bff4bcfa08f119f2227ca5d7d2e3ba623523e219c33728264736f6c63430008110033ccf65503000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000141858c77d000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000014",
  "splittedByteCodes": [
    "610300",
    "6200006c",
    "6080",
    "39",
    "610278",
    "6080",
    "6000",
    "f0",
    "61038052",
    "60ff",
    "60e0",
    "6004",
    "6103a0",
    "6000",
    "6000",
    "6044",
    "6102f8",
    "6000",
    "61038051",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6000",
    "6000",
    "6044",
    "61033c",
    "6000",
    "61038051",
    "5a",
    "f1",
    "84",
    "1b",
    "3d",
    "84",
    "1b",
    "01",
    "81",
    "52",
    "81",
    "01",
    "3d",
    "6000",
    "82",
    "3e",
    "3d",
    "01",
    "6103a0",
    "90",
    "03",
    "6103a0",
    "f3",
    "",
    "608060405234801561001057600080fd5b50610258806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c80631858c77d1461003b578063ccf655031461006b575b600080fd5b61005560048036038101906100509190610102565b61009b565b6040516100629190610151565b60405180910390f35b61008560048036038101906100809190610102565b6100b1565b6040516100929190610151565b60405180910390f35b600081836100a9919061019b565b905092915050565b600081836100bf91906101de565b905092915050565b600080fd5b6000819050919050565b6100df816100cc565b81146100ea57600080fd5b50565b6000813590506100fc816100d6565b92915050565b60008060408385031215610119576101186100c7565b5b6000610127858286016100ed565b9250506020610138858286016100ed565b9150509250929050565b61014b816100cc565b82525050565b60006020820190506101666000830184610142565b92915050565b7f4e487b7100000000000000000000000000000000000000000000000000000000600052601160045260246000fd5b60006101a6826100cc565b91506101b1836100cc565b92508282039050818112600084121682821360008512151617156101d8576101d761016c565b5b92915050565b60006101e9826100cc565b91506101f4836100cc565b92508282019050828112156000831216838212600084121516171561021c5761021b61016c565b5b9291505056fea2646970667358221220f5b5af1db90e0f8d8bff4bcfa08f119f2227ca5d7d2e3ba623523e219c33728264736f6c63430008110033",
    "ccf65503000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000141858c77d000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000014",
  ],
}
`;

exports[`Test predeploy contracts for chain 42161 simple 2`] = `"0x80000020000000000000000000000000000000000000000000000000000000000000001e80000020fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff6"`;

exports[`Test predeploy contracts for chain 42161 simple 3`] = `
[
  {
    "data": "0x000000000000000000000000000000000000000000000000000000000000001e",
    "success": true,
  },
  {
    "data": "0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff6",
    "success": true,
  },
]
`;

exports[`Test predeploy contracts for chain 42161 simple 4`] = `
 [
  30n,
]
`;

exports[`Test predeploy contracts for chain 42161 simple 5`] = `
 [
  -10n,
]
`;
